### ISSUE TITLE
sql: internal executor doesn't have session init'd properly and is just using go defaults

### ISSUE THREAD

**Issue by cucaroach** [cockroachdb/cockroach#70888]
The session data used by the internal executor looks like all the session data defaults are just the go value defaults:

```
SessionData = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.SessionData} 
 Database = {string} ""
 ApplicationName = {string} "$ internal-claim-jobs"
 UserProto = {github.com/cockroachdb/cockroach/pkg/security.SQLUsernameProto} "root"
 DataConversionConfig = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.DataConversionConfig} 
 VectorizeMode = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.VectorizeExecMode} VectorizeUnset (0)
 TestingVectorizeInjectPanics = {bool} false
 DefaultIntSize = {int32} 0
 Location = {string} ""
 SearchPath = {[]string} nil
 TemporarySchemaName = {string} ""
 SeqState = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.SequenceState} 
 WorkMemLimit = {int64} 0
 IntervalStyleEnabled = {bool} false
 DateStyleEnabled = {bool} false
 LockTimeout = {time.Duration} 0 0s
 Internal = {bool} true
 OnUpdateRehomeRowEnabled = {bool} false
LocalOnlySessionData = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.LocalOnlySessionData} 
 SaveTablesPrefix = {string} ""
 OptimizerFKCascadesLimit = {int64} 0
 StmtTimeout = {time.Duration} 0 0s
 IdleInSessionTimeout = {time.Duration} 0 0s
 IdleInTransactionSessionTimeout = {time.Duration} 0 0s
 NoticeDisplaySeverity = {uint32} 0
 ReorderJoinsLimit = {int64} 0
 DefaultTxnPriority = {int64} 0
 DefaultTxnReadOnly = {bool} false
 DefaultTxnUseFollowerReads = {bool} false
 PartiallyDistributedPlansDisabled = {bool} false
 OptimizerUseHistograms = {bool} false
 OptimizerUseMultiColStats = {bool} false
 LocalityOptimizedSearch = {bool} false
 SafeUpdates = {bool} false
 PreferLookupJoinsForFKs = {bool} false
 ZigzagJoinEnabled = {bool} false
 RequireExplicitPrimaryKeys = {bool} false
 ForceSavepointRestart = {bool} false
 AllowPrepareAsOptPlan = {bool} false
 TempTablesEnabled = {bool} false
 ImplicitColumnPartitioningEnabled = {bool} false
 OverrideMultiRegionZoneConfigEnabled = {bool} false
 HashShardedIndexesEnabled = {bool} false
 DisallowFullTableScans = {bool} false
 ImplicitSelectForUpdate = {bool} false
 InsertFastPath = {bool} false
 AlterColumnTypeGeneralEnabled = {bool} false
 SynchronousCommit = {bool} false
 EnableSeqScan = {bool} false
 EnableUniqueWithoutIndexConstraints = {bool} false
 StubCatalogTablesEnabled = {bool} false
 ExperimentalComputedColumnRewrites = {string} ""
 CopyPartitioningWhenDeinterleavingTable = {bool} false
 EnableStreamReplication = {bool} false
 ResultsBufferSize = {int64} 0
 PropagateInputOrdering = {bool} false
 ExperimentalDistSQLPlanningMode = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.ExperimentalDistSQLPlanningMode} ExperimentalDistSQLPlanningOff (0)
 DistSQLMode = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.DistSQLExecMode} DistSQLOff (0)
 SerialNormalizationMode = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.SerialNormalizationMode} SerialUsesRowID (0)
 NewSchemaChangerMode = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.NewSchemaChangerMode} UseNewSchemaChangerOff (0)
 SequenceCache = {github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb.SequenceCache} nil
 PlacementEnabled = {bool} false
 SessionUserProto = {github.com/cockroachdb/cockroach/pkg/security.SQLUsernameProto} ""
 TxnRowsWrittenLog = {int64} 0
 TxnRowsWrittenErr = {int64} 0
 TxnRowsReadLog = {int64} 0
 TxnRowsReadErr = {int64} 0
 AutoRehomingEnabled = {bool} false
 IsSuperuser = {bool} false
 LargeFullScanRows = {float64} 0
```

Is this correct?  Shouldn't we init some values to their cluster setting defaults?   Thinking about EnableZigzagJoin, InsertFastPath, DistSQLMode, VectorizeMode, WorkMemLimit (looks like we interpret 0 as Default, ie 64MB here).    Worried about internal queries running slowly because they aren't using all our chops.

Jira issue: CRDB-10271

**Comment by ajwerner** [comment]
What's the context in which this internal executor is being executed? You're correct that the global one hanging off the execCfg doesn't have any session data set. This came up recently when I was debugging something. 

The values from https://github.com/cockroachdb/cockroach/blob/d88ece55fed5db2f7bdb478254199c2957e13915/pkg/sql/internal.go#L178-L179

If you use the `InternalExecutor` hanging of off the `evalCtx` it should share session settings with the associated connection. Honestly this behavior is also sort of surprising as that's the only state that's really shared between the internal executor and the session. I hate that. See https://github.com/cockroachdb/cockroach/issues/69495 for a different internal executor problem about things not being associated tightly enough with the session.

**Comment by cucaroach** [comment]
> What's the context in which this internal executor is being executed? 

I didn't catalog them all but I think it was mostly jobs related?

**Comment by ajwerner** [comment]
> > What's the context in which this internal executor is being executed? 
> 
> 
> 
> I didn't catalog them all but I think it was mostly jobs related?

Yeah, I jobs were not doing anything with session state. My guess is that we should rework the "default" session state for the global internal executor. I think it's a pretty small change but I suspect there'll be some surprising fallout.

**Comment by cucaroach** [comment]
To be exact I was going around in circles trying to figure out why EnableZigzagJoin was false.    This is purely about initializing sessions with the correct default state which isn't necessarily the same as Go default state which seems like an accidental oversight.

**Comment by ajwerner** [comment]
> To be exact I was going around in circles trying to figure out why EnableZigzagJoin was false. This is purely about initializing sessions with the correct default state which isn't necessarily the same as Go default state which seems like an accidental oversight.

Right, agreed. I ran into this the other day too doing something similar ([slack thread](https://cockroachlabs.slack.com/archives/C01RX2G8LT1/p1631574901092000)). Very painful. 

I'm just saying the fix is to change the logic here. where we construct the session state if it's not provided (linked above and again below). I speculate that it'll break something subtle that's relying on these zero values.

https://github.com/cockroachdb/cockroach/blob/d88ece55fed5db2f7bdb478254199c2957e13915/pkg/sql/internal.go#L178-L179

**Comment by rafiss** [comment]
I think https://github.com/cockroachdb/cockroach/pull/71246 might fix this -- is that right @RichardJCai ?

**Comment by RichardJCai** [comment]
> I think #71246 might fix this -- is that right @RichardJCai ?

Yeah I believe it should since we now have to pass session data to create the InternalExecutor.

**Comment by cucaroach** [comment]
Woot, thanks, I'll verify and close this when that PR lands.

**Comment by rafiss** [comment]
Hm nevermind, I might have been mistaken. There is now a `InternalExecutorFactory` as a field of `ExecCfg`, but there still is also the `ExecCfg.InternalExecutor`, which has ~158 usages and still has this issue. So I believe the work for this issue would be to move all those usages to use the `InternalExecutorFactory` instead.

**Comment by vy-ton** [comment]
@rafiss This came up in Queries triage/discussion. I wasn't sure if there were plans to tackle it with https://cockroachlabs.atlassian.net/browse/CRDB-14492?

**Comment by ajwerner** [comment]
cc @ZhouXing19 the (perhaps reluctantly) up and coming internal executor expert.

**Comment by cucaroach** [comment]
I gave this another go on Friday and didn't get super far but regardless happy to collab or handoff.  Probably one of those things better to get into 22.2 early to weed out all the gremlins.

**Comment by ZhouXing19** [comment]
I'll pick this up, maybe along with https://github.com/cockroachdb/cockroach/pull/80262, which is refactoring the internal executor initialization.

**Comment by cucaroach** [comment]
It would be nice to see where the go defaults and the cluster defaults diverge, it would be a much smaller list.  The problem is not all internal executors are the same and there's specific overrides at play (like migration jobs opt out of distsql).  The crime is things like InsertFastPath are disabled.

**Comment by cucaroach** [comment]
Although you do bring up a good point about cluster settings that have side affects, like if sql.trace.txn.threshold is on should that apply to internal executors?

**Comment by ZhouXing19** [comment]
> @ZhouXing19 When you take a look here, could you compile a list of settings where the go default differs from the CockroachDB default?

Made a list for entries in `SessionData` and `LocalOnlySessionData`: https://docs.google.com/spreadsheets/d/1QgwM8qHJtRsPApy0LCHw_nPvg_xZi4VM1EwAiHCQ0V0/edit?usp=sharing
cc @vy-ton

**Comment by ajwerner** [comment]
I keep stumbling over this. I added a new feature and I started with a cluster settings. Then I switched it to use a session variable and I almost thought I had done it right, but noticed that the CPU usage of background internal queries was still high. It's awkward because I don't know that I feel comfortable setting my session variable up as part of the ["minimal session data"](https://github.com/cockroachdb/cockroach/blob/d1962deb662c115ceec158a0c09d3f797eca3e71/pkg/sql/conn_executor.go#L875).
